### ISSUE TITLE
Properly set object status

### ISSUE THREAD

**Issue by ddysher** [kubernetes/kubernetes#2726]
As per discussion https://github.com/GoogleCloudPlatform/kubernetes/pull/2665#discussion_r21126784, we currently haven't set object status for watch command.  The status should be set properly before returning.

**Comment by bgrant0607** [comment]
@smarterclayton Annotations are intended to record management/deployment information not understood by the core system. By definition, anything written to Status by a controller is understood by the core implementation of that controller and so doesn't belong in annotations. Anything a controller has to report should go into Status, Events, a synthetic view object, like Endpoints, or a synthetic control resource, like Bindings. Some controllers also mutate the desired state of objects, as in the case of creating pods or resizing a replicationController, but I think that's likely to be an independent process/datapath.

**Comment by smarterclayton** [comment]
But are there external systems that want to mutate status (unhealthy) *and* add annotations that record additional status not captured in the core data for themselves to use later?  I.e. they might want to create their own resource "foo" which results in the status of the node getting updated, and when they update the status they want to atomically record on the node what the resource version of "foo" they were based on?

----- Original Message -----

**Comment by derekwaynecarr** [comment]
Is this a v1beta3 or something after timeframe?  The discussion here seems so fundamental and has so many ramifications depending on the route taken.  Another question, is kubectl a valid client for a controller in your view?  Meaning it too should be able to update Status.  I hope answer is yes, but wanted to be sure that we were not making an assertion as to how a controller is written by others.

**Comment by smarterclayton** [comment]
We either have to restrict object mutation via a new resource (binding) or field level authz (finalizers).  There are advantages and disadvantages to both, and we could realistically do simple things now that allow mutation on some resources (nodes and your stuff).  I think we have enough info in various places to set a path, but we need to get it on paper.

**Comment by bgrant0607** [comment]
Preventing accidental user clobbering of system-set status (#3005) is a requirement for 1.0. The rest isn't, unless mandated by performance concerns or needed to alleviate complexity in specific components.

**Comment by bgrant0607** [comment]
@smarterclayton Example of an automation system wanting to add annotations: https://github.com/GoogleCloudPlatform/kubernetes/issues/2483#issuecomment-72777069

**Comment by ddysher** [comment]
Start thinking about node decommission, one use case for annotations might be the grace period for a node to be considered terminated.  We keep an integer annotation in node controller, increase the count in each sync period if node check failed, decrease the count/or reset to 0 when check succeeds.

We can start with an in-memory solution, but eventually it's better to persist it with node status.

**Comment by bgrant0607** [comment]
The "pod cache" is such a nightmare:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/master/pod_cache.go

See #4169, #4160, #4166, #1366, and others.

**Comment by bgrant0607** [comment]
I'd like to convert this from a design issue to an implementation issue.

Strawman proposal:

Like Binding and the proposed Stop #1535, Resize #1629, and Clone, I propose a Custom Verb, aka Synthetic Control Object, tentatively called UpdateStatus, though it would also allow setting annotations and possibly labels, too, as was requested for Binding (#4103). kubectl would get a new updatestatus command, which would PUT to the UpdateStatus resource corresponding to any object. 

As for the path, as discussed in #3652, I propose a subpath under the object, using a collection name that isn't otherwise a valid object name. I previously proposed `~`, but actually `!` is [apparently valid](http://stackoverflow.com/questions/4669692/valid-characters-for-directory-part-of-a-url-for-short-links), and is suggestive of an action. For example: `pods/foopod/!/updatestatus`.

I don't think we want the behavior to be implicitly determined by auth, and we've shot down URL parameters in other discussions (e.g., #148, #170).

**Comment by bgrant0607** [comment]
While we're at it, I propose changing Binding to use the same path format, whatever it is.

**Comment by smarterclayton** [comment]
The alternate path proposal would be `PUT /objectstatus/pods/foopod` - it means we potentially constrain the space of future resources so that no one can ever have a 'objectstatus' resource, so I don't think it's as good.

"updatestatus" is a bit of an RPCism - the restful-approach might be `PUT /pods/foopod/!/status`.  Do we need the `/!/` in the middle?  I don't think we reserve anything below `/<resource>/<name>/`, unless we want to introduce further path segments in the name (unlikely), so `/<resource>/<name>/<subresource>` isn't unreasonable.

**Comment by smarterclayton** [comment]
Suggestions for proposed verbs:

```
POST /pods/mine/stop    # generic verb
PUT /pods/mine/resize   # with resize object if using absolute, POST if relative
POST /pods/mine/clone  # with Location header
```

**Comment by derekwaynecarr** [comment]
+1 to subpath and not using verb in path

**Comment by bgrant0607** [comment]
I see the point about avoiding RPCisms, but stop, resize, and clone are also verbs. We could nounify them: stopper, resizer, cloner?

The reason for `!` is to facilitate querying for what subpaths exist, in the REST collection/instance style. We could use a plural noun, but I'm not sure what it would be if we don't want to consider these to be verbs. Previously I've proposed `ops` and `methods`. Maybe `controls`.

**Comment by bgrant0607** [comment]
Or `views`.

**Comment by smarterclayton** [comment]
Ah - it's not something I've seen commonly (or effectively) done in the past, so was not at the top of my mind.  Most times people solve it with url templates on the root resource (you can perform the X action on the Y resource by templating this url form).  It's not a bad pattern, although not in common use so might confuse some folks.

status and bindings are definitely a resource if we want to GET and PUT it.  resize is can be seen as a "size" resource, where you GET / PUT / POST to mutate a set of scale parameters (it coincidentally gives you an easy way to see the current scale via get).

I agree clone has no clear resource analogue unless you want to do dry runs of clone, so a control verb is not unwarranted.  I typically assume one off actions are best represented as POST to a well named endpoint in REST, so as long as the docs are good and the name is well chosen I don't think it deserves sophistication.

I'll go on record as not being a die hard rest fanatic, just that anything with a read / write cycle generally can be a resource (and is easy to write clients for), and we've stuck with the pattern so far.  I'm fine with adding the extra level to allow easy introspection, although we don't have to force it to be a parent directory of the other resources.

**Comment by smarterclayton** [comment]
One question is whether we intend to allow the actions to vary among a single resource type, which I assume we do not, so the /! endpoint is purely informational, and could be at a higher level.  However, people may assume /! Is available or not based on whether the named resource exists, and make assumptions (incorrectly) the wrong way. If it's a discovery endpoint that clients cache or query, it might be easier to place it on the root api document (/api/v1beta1) so they can get everything at once

**Comment by bgrant0607** [comment]
Yeah, GCE (not that it's the best API model) also doesn't put methods into a collection. For example:
https://cloud.google.com/compute/docs/reference/latest/instances/attachDisk

I suppose I could live with that. We also have the swagger docs for the paths.

I'm fine with status, binding, and size. Clone is a noun as well as a verb, so that could slide. Stop perhaps could be `end`. Or we could generalize it to `control` (as in the noun), for all common controls, including graceful termination but maybe also activation or suspension in the future.

**Comment by bgrant0607** [comment]
Moved clone discussion back to #170.

Another round on the other paths:
```
/pods/mine/control # general path for uniform object controls, such as stop, and probably others
/pods/mine/size
/pods/mine/status
/pods/mine/binding
```

One thing mentioned this morning is whether the status path should also contain the spec: I'd like the status path to use a different object type so that we could label the status field as readonly in the user-oriented type, so I'm thinking status would contain metadata and status, but not spec. I'd expect the component to watch the parent path and just write to the subpath.

**Comment by smarterclayton** [comment]
> On Feb 6, 2015, at 4:14 PM, Brian Grant <notifications@github.com> wrote:
> 
> Moved clone discussion back to #170.
> 
> Another round on the other paths:
> 
> /pods/mine/control # general path for uniform object controls, such as stop, and probably others
> /pods/mine/size
> /pods/mine/status
> /pods/mine/binding
> One thing mentioned this morning is whether the status path should also contain the spec: I'd like the status path to use a different object type so that we could label the status field as readonly in the user-oriented type, so I'm thinking status would contain metadata and status, but not spec. I'd expect the component to watch the parent path and just write to the subpath.
> 
Agree - will look at implementation early next week.

**Comment by bgrant0607** [comment]
Awesome, thanks.

**Comment by smarterclayton** [comment]
#4248 is where I'm going to do this.  To properly handle a sub resource collection, I'm going to strip away some layers of RESTHandler that are no longer needed (now that nikhil has added api_installer, and Derek brought in proper path handling).  That "unwrapping" will also be necessary for authz to identify the resources.

@deads2k we need to chat today about how to represent sub resource mutation via policy rules in this context - setting the status of a pod is a separate authorization (that most users won't have), but it's not quite a separate resource.

**Comment by bgrant0607** [comment]
We should ensure that watching status works once we've switched to this new mechanism.

**Comment by davidopp** [comment]
Can I assign this to @smarterclayton? IIUC this is considered a blocker for 1.0, and we should have every 1.0 issue assigned to someone.

**Comment by smarterclayton** [comment]
Yes

**Comment by bgrant0607** [comment]
The spec and status will be updated by different entities, at different times.

Users, rollout tools, and automated services (auto-scalers, etc.) will update metadata and spec relatively frequently, but probably not continuously. Still, once auto-scaling and auto-sizing are common, safe updates (i.e., using resourceVersion preconditions) initiated by users will often need to be retried before they can succeed. We should ensure that our client can do that robustly and cleanly. However, if the user had to turn off automatic updates, they should be able to do that (e.g., by disabling auto-scaling).

Controller components will update status continuously, relentlessly. Users should have no ability to stop this. This will continuously change resourceVersion and, once we add it, lastModificationTime.

These really feel like different objects stored in the same record, thus producing false sharing and creating merge complexity.

Returning a unified object on GET is nice from a UX perspective, but unified objects increase complexity for virtually all writes, for both API clients and for the API implementation.

Once we have multi-object transactions (at least for creation and deletion), we should seriously consider storing metadata+spec and (partial metadata+)status in separate records. 

Regardless of the underlying implementation, I think a subresource is the right way to think about status, as opposed to a special case of patch.

**Comment by smarterclayton** [comment]
> On Feb 23, 2015, at 4:54 PM, Brian Grant <notifications@github.com> wrote:
> 
> The spec and status will be updated by different entities, at different times.
> 
> Users, rollout tools, and automated services (auto-scalers, etc.) will update metadata and spec relatively frequently, but probably not continuously. Still, once auto-scaling and auto-sizing are common, safe updates (i.e., using resourceVersion preconditions) initiated by users will often need to be retried before they can succeed. We should ensure that our client can do that robustly and cleanly. However, if the user had to turn off automatic updates, they should be able to do that (e.g., by disabling auto-scaling).
> 
> Controller components will update status continuously, relentlessly. Users should have no ability to stop this. This will continuously change resourceVersion and, once we add it, lastModificationTime.
> 
A good point, and has the potential to drive the design more than others.  However, what are the changes that matter, and will they result in deltas?  Restart time, attempts, should converge to one write per change - the simpler an object's status the less likely is a write.

However, user intent changes and object status changes do have different retention rates (although in a perfect world you'd capture both).  If we don't store them in the same object, I think we'd want to ensure watch works on objects for status.  Most of the watch use cases are reactive (when X is ready, when Y dies).

**Comment by bgrant0607** [comment]
I totally agree that we need to ensure that watch works for status.

**Comment by smarterclayton** [comment]
Minimal change for status is up - there's some factoring that I want to do, but it's working

**Comment by smarterclayton** [comment]
Next step after #4779 is to implement a controller that sets PodStatus and disables the pod cache decorator.  Need to test how many writes to etcd this adds in our core create path.

**Comment by bgrant0607** [comment]
Or skip the controller and do #4561.

**Comment by smarterclayton** [comment]
Better

**Comment by bgrant0607** [comment]
Can anyone think of a reason why we should initialize status in the defaulting pass? We'll need to initialize PodPhase to Pending, for example.

**Comment by fgrzadkowski** [comment]
@bgrant0607 @smarterclayton I'm already working on #4561. Waiting for #4779 to be merged.

**Comment by smarterclayton** [comment]
Great

**Comment by lavalamp** [comment]
@bgrant0607 can we change this title to "Add subresource types (/status etc)"?

**Comment by bgrant0607** [comment]
@lavalamp We should probably file other issues re. the other subresources.

**Comment by smarterclayton** [comment]
I'm currently working on clone/instantiate and getting status cleaned up for other resources.

**Comment by bgrant0607** [comment]
Filed #5000 for /binding. There are already issues for /size and stop, and this issue is about /status. If we want to do something similar for /events and /endpoints, those would be 2 new issues.

**Comment by bgrant0607** [comment]
Cool. Clone/instantiate is covered by #170.

**Comment by davidopp** [comment]
Can someone (maybe Clayton?) summarize the status of this issue? If this has become a meta-issue, let's file any necessary sub-issues and close this. Otherwise, what remains to be done here for 1.0?

**Comment by smarterclayton** [comment]
I don't know that we have a meta issue covering "cluster mutation is watchable, and the status of the cluster is observable from the API server without passthrough hacks" except this one.  Will try to summarize if brian doesn't get here first.

**Comment by bgrant0607** [comment]
Quick status AIUI:
* Subresource paths are now supported by the apiserver library
* Binding has been converted to a subresource
* Pod status is posted from Kubelet to apiserver via a subresource
* Node status is posted, but not via a subresource AFAIK; Node is being ported to generic etcd, though #5441
* Replication controller status is posted, but not via a subresource #4909
* Service doesn't yet have status
* Objects whose status is posted should be watchable
* pod_cache deletion PR is awaiting merge #5854
* We decided to use DeleteOptions rather than /stop
* For now we're using a PodTemplate (#5012) resource rather than /clone
* I don't think any work has been done on /size, but that's fairly off topic

**Comment by smarterclayton** [comment]
PodTemplate would support /instantiate.

I need to do some prep work for /size to allow easy implementation.

----- Original Message -----

**Comment by derekwaynecarr** [comment]
Namespace Status is posted via a subresource

**Comment by davidopp** [comment]
Thanks for the summary. So, concretely, what remains as a 1.0 blocker? Just the things from Brian's list minus the entries that say the item is done?

**Comment by bgrant0607** [comment]
I had hoped that all status would be posted via subresources in the v1 API. Right now the API is in an intermediate state. I think finishing the remaining resources is mainly additive and therefore wouldn't need to be done for 1.0. Whether via a subresource or via posting the full resource is mostly an implementation detail. What matters for clients is the ability to watch status changes, which just requires status to be posted by some means.

Remaining things to be fixed for 1.0 can be driven by usability and performance problems.

**Comment by bgrant0607** [comment]
Using more specific issues for the remainder of the work.
